### PR TITLE
Create TPM eventlog entries from measure-config measurements

### PR DESCRIPTION
**Pull request by shjala**
This PR cleans up and simplifies some existing TPM debugging mechanism and adds a new one to measure-config. 

Measure-config measures `/config` content (with some exceptions) into PCR-14, but because it is happening outside of normal TPM measured boot process, no event log entries are added the `binary_bios_measurements`. Changes in this PR enable measure-config to create event log entries with TCG specified format and dumped them into a file, this file is later picked up by `tpm.go`, and its content gets appended to the `binary_bios_measurements` to create the final TPM event log. This can help us debug vault unlock issues that arise from PCR mismatches with PCR-14 being the faulty index.

Example TPM final log created by EVE and successfully parsed by `tpm2_eventlog`, see the entries after `EventNum: 200` : 
 
```
~ tpm2 eventlog tpm_measurement_seal_success
---
version: 1
events:
- EventNum: 0
  PCRIndex: 0
  EventType: EV_NO_ACTION
  Digest: "0000000000000000000000000000000000000000"
  EventSize: 45
  SpecID:
  - Signature: Spec ID Event03
    platformClass: 0
    specVersionMinor: 0
    specVersionMajor: 2
    specErrata: 0
    uintnSize: 2
    numberOfAlgorithms: 4
    Algorithms:
    - Algorithm[0]:
      algorithmId: sha1
      digestSize: 20
    - Algorithm[1]:
      algorithmId: sha256
      digestSize: 32
    - Algorithm[2]:
--More--WARN: Event 55 is unexectedly not extending either PCR 8, 9, 12 or 14
      algorithmId: sha384
      digestSize: 48
    - Algorithm[3]:
      algorithmId: sha512
      digestSize: 64
    vendorInfoSize: 0
- EventNum: 1
  PCRIndex: 0
  EventType: EV_S_CRTM_VERSION
  DigestCount: 4
  Digests:
  - AlgorithmId: sha1
    Digest: "1489f923c4dca729178b3e3233458550d8dddf29"
  - AlgorithmId: sha256
    Digest: "96a296d224f285c67bee93c30f8a309157f0daa35dc5b87e410b78630a09cfc7"
  - AlgorithmId: sha384
    Digest: "1dd6f7b457ad880d840d41c961283bab688e94e4b59359ea45686581e90feccea3c624b1226113f824f315eb60ae0a7c"
  - AlgorithmId: sha512
    Digest: "5ea71dc6d0b4f57bf39aadd07c208c35f06cd2bac5fde210397f70de11d439c62ec1cdf3183758865fd387fcea0bada2f6c37a4a17851dd1d78fefe6f204ee54"
  EventSize: 2
  Event: "0000"
- EventNum: 2
  PCRIndex: 0
  EventType: EV_EFI_PLATFORM_FIRMWARE_BLOB
  DigestCount: 4
  Digests:
  - AlgorithmId: sha1
    Digest: "b7307b419107c2b6c0af7f3f034d30071c52865f"
  - AlgorithmId: sha256
    Digest: "bce983ca294be2a4419926d83ca2ca4f75fca6bad44a1e62a53a5722577f7f09"
  - AlgorithmId: sha384
    Digest: "fad334d26232bafdc251fbde878a7e3bdd03a672fd4ffc2568b28c71ce88abd442aebafd2d2c78efeec366a105cd9479"
  - AlgorithmId: sha512
    Digest: "e481240bf911d94df4a7608be06bbc2c8a2ac93fe115356527b5b81c398c12a3c1787f2349bdaea5e27be8be62205c7a814691a7d81447408e1f37d084c74f37"
  EventSize: 16
  Event:
    BlobBase: 0x1000
    BlobLength: 0x1ff000
[...]
- EventNum: 200
  PCRIndex: 5
  EventType: EV_EFI_ACTION
  DigestCount: 4
  Digests:
  - AlgorithmId: sha1
    Digest: "475545ddc978d7bfd036facc7e2e987f48189f0d"
  - AlgorithmId: sha256
    Digest: "b54f7542cbd872a81a9d9dea839b2b8d747c7ebd5ea6615c40f42f44a6dbeba0"
  - AlgorithmId: sha384
    Digest: "0a2e01c85deae718a530ad8c6d20a84009babe6c8989269e950d8cf440c6e997695e64d455c4174a652cd080f6230b74"
  - AlgorithmId: sha512
    Digest: "1bb30cdbd6da78fe2a8a161ef51176e22d64dce305b40b47243673af64a2b16fca6182116433e3891be94773f6d7d411275721d5bf7d40ea51a274d5c891637c"
  EventSize: 40
  Event: |-
    Exit Boot Services Returned with Success
- EventNum: 201
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "09314b1d577540a6316cb49761ebe023233b3b28aed683a1c5bc688c6cdf9892"
  EventSize: 55
  Event: |-
    file:/config/DevicePortConfig/override.json exist:false
- EventNum: 202
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "647216e535c2e8f291d96bf53edcf014b6acb89ee97ae6125ed4d167847e8437"
  EventSize: 141
  Event: |-
    file:/config/DevicePortConfig/override.json.template exist:true content-hash:6b2f947e3071058d57f8093229b668c630f4eb7a11066c498e6dec6234bf1d68
- EventNum: 203
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "4ea84f2d1abf07bd87f440b610c063b37bbfa4a08b1e0cd08b1ab284c20668b9"
  EventSize: 37
  Event: |-
    file:/config/Force-API-V1 exist:false
- EventNum: 204
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "4d4e040899791e1272a878347e885b5a88c3950edf8bb09cccdae7ff8735e6d5"
  EventSize: 49
  Event: |-
    file:/config/GlobalConfig/global.json exist:false
- EventNum: 205
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "ca7c8cca41b6a16477a53e1464a6297152ecb19ae40d07cef02121577c52de87"
  EventSize: 44
  Event: |-
    file:/config/bootstrap-config.pb exist:false
- EventNum: 206
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "2124645fdd1b2f3be5e66bdaea35eec70d4c497201f6b99acebd8c667ce667f0"
  EventSize: 40
  Event: |-
    file:/config/device.cert.pem exist:false
- EventNum: 207
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "7c57a2b8790a12129f900f18409d3bd900611b411c8d56a17723c6920ad03c47"
  EventSize: 39
  Event: |-
    file:/config/device.key.pem exist:false
- EventNum: 208
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "e29b5b4e1b2545107cebab84645701cb5120faacbfb51934d6b1a9f962946559"
  EventSize: 110
  Event: |-
    file:/config/grub.cfg exist:true content-hash:d462db1231cd467330fdb672934262f627566bbda2c1805289e35d118f13f581
- EventNum: 209
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "3dbf42deeb4d9d012358ed68e009bb653335efe050f8872ad1ba52577861d671"
  EventSize: 115
  Event: |-
    file:/config/grub.cfg.tmpl exist:true content-hash:d462db1231cd467330fdb672934262f627566bbda2c1805289e35d118f13f581
- EventNum: 210
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "8de516ce11a04cb77f62461917cbaa47d8add45afb52328a614d525a7c23dd29"
  EventSize: 40
  Event: |-
    file:/config/onboard.cert.pem exist:true
- EventNum: 211
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "1a112c245a8c3ae59dadee9989662bced2ad93c578df5c115d740ea86e65d35d"
  EventSize: 39
  Event: |-
    file:/config/onboard.key.pem exist:true
- EventNum: 212
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "16c616c536e161e76a0e2ab4ffaa181dd3eb7f76082bdd727b4e8fcd437fcdfc"
  EventSize: 169
  Event: |-
    file:/config/origin.2023-11-20.10.4.1-12-gbf81b337c-20-g0c123dcd5-294-g816d53495 exist:true content-hash:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
- EventNum: 213
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "50b21c9dbcf8a1358242e5126b9384997a834e14ef27a081da2cbe504e54a591"
  EventSize: 122
  Event: |-
    file:/config/root-certificate.pem exist:true content-hash:a24d486bf6381401ebb1120d6df537bd082bacd5afae02ab2f503c2380b8a341
- EventNum: 214
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "78336fe76a5a814f0a5f727b32edd2b1f9cb544777c76ae94852f4045751326c"
  EventSize: 108
  Event: |-
    file:/config/server exist:true content-hash:0a24b79ed482d60e95a9d1f5f2ed32cc6cf18fc13af51c1c66bc6c4de0d57939
- EventNum: 215
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "0928fa49f626fa362fcf4d41c021822cb3eaff0eeb187a754950ce8bdf4ff29a"
  EventSize: 36
  Event: |-
    file:/config/soft_serial exist:false
- EventNum: 216
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "1dd5441be318ec22f21d2ed14202cd94c9513a0edd00dbb64211ffe13256d73d"
  EventSize: 39
  Event: |-
    file:/config/tpm_credential exist:false
- EventNum: 217
  PCRIndex: 14
  EventType: EV_EFI_ACTION
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "062d46f4d0ad791ae5240541b54d45176e68f09d5a89c98beb61fdfb6a5e2998"
  EventSize: 132
  Event: |-
    file:/config/v2tlsbaseroot-certificates.pem exist:true content-hash:bb1451b8335cd0ef0f8d6e515154c94d764f1ddd0b247f5e6199ae3b2deec930
```

### CODE REVIEW

**Comment by shjala** [comment]
I'll fix yetus complaints once PR is reviewed.

**Review by eriknordmark — COMMENTED** [review]
Two comments/questions.

File: pkg/measure-config/src/measurefs.go [hunk 1]
```diff
@@ -95,34 +201,55 @@ func sha256sumForFile(filePath string) (string, error) {
...
-		length = min(maxEventDataSize, len(eventData)-offset)
-		// PCREvent internally hashes the data with all supported algorithms
-		// associated with the PCR banks, and extends them all before return.
-		err := tpm2.PCREvent(tpm, configPCRHandle, []byte(eventData[offset:offset+length]))
-		if err != nil {
-			return nil, fmt.Errorf("cannot measure %s. couldn't extend PCR: %v", filePath, err)
-		}
+	// PCREvent internally hashes the data with all supported algorithms
+	// associated with the PCR banks, and extends them all before return.
+	err := tpm2.PCREvent(tpm, configPCRHandle, []byte(eventData))
```
[the remark this answers is not in the document]
Comment (shjala) [reply pkg/measure-config/src/measurefs.go:225]: No, we still calculate it the same, see [L114](https://github.com/lf-edge/eve/pull/3607/files/816d5349569950f0d3345eb21cf0b867c567d39b#diff-4cccad80ff862a963fd4b5f8bb1dfc0d83081f8ad3230a7ff3e8ffa920defec5L114), and Event Data (that is hashed by TPM) is formed from  this format string`file:%s exist:true content-hash:%s` which never reaches 1024 bytes in reality, the removed loop at [L108](https://github.com/lf-edge/eve/pull/3607/files/816d5349569950f0d3345eb21cf0b867c567d39b#diff-4cccad80ff862a963fd4b5f8bb1dfc0d83081f8ad3230a7ff3e8ffa920defec5L108) to break data into chunks if > 1024, is basically a dead code.

[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/measure-config/src/measurefs.go:225]: OK

File: pkg/measure-config/src/measurefs.go [hunk 2]
```diff
@@ -226,14 +353,37 @@ func readConfigPCR(tpm io.ReadWriter) ([]byte, error) {
 func main() {
 	tpm, err := tpm2.OpenTPM(TpmDevicePath)
 	if err != nil {
-		log.Printf("couldn't open TPM device %s. Exiting", TpmDevicePath)
-		return
+		log.Fatalf("couldn't open TPM device %s. Exiting", TpmDevicePath)
```
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/measure-config/src/measurefs.go:356]: OK, then we'll depend on the regular periodic tests of master for this.

**Review by eriknordmark — APPROVED** [review]
LGTM

**Comment by shjala** [comment]
fixed yetus, broke the last commit in two to make it easier to backport it to LTS.

**Review by eriknordmark — APPROVED** [review]
LGTM

**Pull request merged**